http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Http2IT.java

```diff
@@ -0,0 +1,261 @@
+        httpClient = vertx.httpClientBuilder().with(options).build();
+        httpClient.request(HttpMethod.GET, app.getPort(Protocol.HTTPS), baseUri.getHost(), BASE_ENDPOINT)
+                .compose(request -> request.send()
+                        .compose(httpClientResponse -> {
+                            assertEquals(HttpStatus.SC_OK, httpClientResponse.statusCode());
+                            String contentTypeOptionsHeader = httpClientResponse.getHeader("X-Content-Type-Options");
+                            assertNotNull(contentTypeOptionsHeader);
+                            assertEquals("nosniff", contentTypeOptionsHeader);
+                            assertEquals(HttpVersion.HTTP_2, httpClientResponse.version());
+                            checkpoint.flag();
+                            return httpClientResponse.body();
+                        }))
+                .onSuccess(body -> {
+                    assertThat("Body response", body.toString().contains(GREETING));
+                    checkpoint.flag();
+                })
+                .onFailure(vertxTestContext::failNow);
+    }
+
+    @Test
+    @Order(6)
+    @DisplayName("Test sending and receiving custom frames over HTTP/2")
+    void customFramesTest(VertxTestContext context) {
+        AtomicLong timerId = new AtomicLong();
+        Checkpoint entry = context.checkpoint(2);
+        HttpClientOptions options = new HttpClientOptions().setSsl(true).setUseAlpn(true).setProtocolVersion(HttpVersion.HTTP_2)
+                .setTrustOptions(new JksOptions().setPath("keystore.jks").setPassword("password"))
+                .setVerifyHost(false);
+        httpClient = vertx.httpClientBuilder().with(options).build();
+        httpClient.request(HttpMethod.GET, app.getPort(Protocol.HTTPS), baseUri.getHost(), "/ping")
+                .onFailure(context::failNow)
+                .flatMap(request -> request.sendHead().map(head -> request))
+                .map(request -> {
+                    Future<HttpClientResponse> pong = request.response().onSuccess(response -> {
+                        response.customFrameHandler(frame -> {
+                            context.verify(() -> {
+                                assertEquals(PING_DATA, frame.payload().toString().toLowerCase());
+                                entry.flag();
+                            });
+                        });
+                    });
+                    request.exceptionHandler(context::failNow);
+                    vertx.setPeriodic(1000, timerID -> {
+                        timerId.set(timerID);
+                        request.writeCustomFrame(10, 0, Buffer.buffer("ping"));
+                        entry.flag();
+                    });
+                    return pong;
+                })
+                .result();
+    }
+
+    @AfterEach
+    public void closeConnections() {
+        httpClient.close();
+        vertx.close();
```

Comment:
both returns future, IMO you should make sure that process has finished before you continue. thoughts?

Reply:
From the vert.x docs:

https://vertx.io/docs/vertx-junit5/java/#_closing_and_removal_of_vertx_objects 
_"Injected Vertx objects are being automatically closed and removed from their corresponding scopes.
Similarly, when it is being created by a @BeforeEach method, it is being closed after possible @AfterEach methods have completed."_
https://vertx.io/docs/vertx-core/java/#_closing_the_underlying_connection
_you can [close](https://vertx.io/docs/apidocs/io/vertx/core/http/HttpServerResponse.html#close--) the underlying TCP connection with
close.
Non keep-alive connections will be automatically closed by Vert.x when the response is ended._

Reply:
> Yes returns Future but if I don't recall wrong vert.x framework already managed properly those instances (vert.x and httpclient) and is no need to close it manually.

I certainly know nothing about this framework, but it is highly suspicious from reading docs because:
1. https://vertx.io/docs/vertx-junit5/java/#_closing_and_removal_of_vertx_objects speaks about injected (you don't do that)
2. https://vertx.io/docs/vertx-junit5/java/#_closing_and_removal_of_vertx_objects also speaks about `Similarly, when it is being created by a @BeforeEach method` but it is unclear whether they mean 
```
@BeforeEach
  @DisplayName("Deploy a verticle")
  void prepare(Vertx vertx, VertxTestContext testContext) {
    vertx.deployVerticle(new SomeVerticle(), testContext.succeedingThenComplete());
  }
```
or what you do (`Vertx.vertx()`)
4. if I am to believe they are able to automatically close `Vertx.vertx()` how they exactly know it was created and how they access it? I suppose there are some events or do they bytecode transform code or something?

If you are 100 % sure, then alright. Let's keep it.

Reply:
Still, I do not doubt you know more about this and I'm sorry for digging into it but here https://vertx.io/docs/vertx-junit5/java/#_running_tests_on_a_vert_x_context I can see that

1. they are not using static field for the `Vertx` instance. is that significant? should we do the same? I think it might because they don't use the `Vertx` instance, so it looks like they can access it.
2. they create Vertx with `vertx = rtoc.vertx();` is that preferred way? I can see that in the very first test they created with `Vertx.vertx()` and didn't care about closing so it signal it is alright.

So here is my suggestion:

- don't use static field or please give me something I can read up about why it's fine

tbh I'd be calmer if you injected the Vert.x, but you are the expert here so I'll trust you.

Reply:
Yes, I agree with injecting Vert.x in this test context. I've done that now.